client/src/rpc_client.rs

```diff
@@ -128,16 +130,51 @@ impl RpcClient {
         Self::new_with_timeout(url, timeout)
     }
 
+    pub fn commitment(&self) -> CommitmentConfig {
+        self.commitment_config
+    }
+
+    fn use_deprecated_commitment(&self) -> Result<bool, RpcError> {
+        let use_deprecated_commitment = self.use_deprecated_commitment.read().unwrap();
+        if let Some(use_deprecated_commitment) = *use_deprecated_commitment {
+            Ok(use_deprecated_commitment)
+        } else {
+            drop(use_deprecated_commitment);
+            let cluster_version = self.get_version().map_err(|e| {
+                RpcError::RpcRequestError(format!("cluster version query failed: {}", e))
+            })?;
+            let cluster_version =
+                semver::Version::parse(&cluster_version.solana_core).map_err(|e| {
+                    RpcError::RpcRequestError(format!("failed to parse cluster version: {}", e))
+                })?;
+            let use_deprecated_commitment = cluster_version < semver::Version::new(1, 5, 5);
+            *self.use_deprecated_commitment.write().unwrap() = Some(use_deprecated_commitment);
```

Comment:
I know this is in the tx encoding check as well and I don't even recall why the `RwLock` was necessary, but I feel like we should hold the write lock _before_ calling `get_version()` to prevent simultaneous callers making the RPC request.

Reply:
Done in https://github.com/solana-labs/solana/pull/14797/commits/b39aa44cdaf2197b223804b85f2693fce6f55b6c